qa: Improve teuthology test failure reason when daemon crashes

bill-scales opened the pull request:
Improvement to the daemon watchdog that monitors daemons and fails the test if a daemon has stopped responding to provide better information about the daemons that have stopped responding, including an assert string and a condensed backtrace if a daemon has crashed.

https://pulpito.ceph.com/billscales-2025-08-15_13:42:55-rados:thrash-erasure-code-optimizations-main-distro-default-smithi/ is an example of a failed test with these changes. Prior to pull request 64889 this would have been reported as a dead job (and taken 8 hours to fail). With pull request 64889 it would have been reported as a watchdog timeout and failed a lot quicker.

The formatting of the job failure needs tweaking to preserve whitespace and newlines - this requires a tweak to the CSS files for pulpito and pulpito-ng. Separate PRs will deliver these changes.

Here is what it looks like with a modified CSS with some annotations to show the various techniques used to condense the failure reason (the first attempt at this PR produced a 30 line failure reason which is far too much detail). There is an unnecessary blank line at the start of the failure reason - this is also removed with the pulpito change

<img width="1265" height="649" alt="teuthologycrash" src="https://github.com/user-attachments/assets/2817b1f8-0819-4390-994a-73830936a77e" />

The objective of the change is to provide enough information in the failure reason for daemon crashes to be able to search for duplicates in the tracker.
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

bill-scales commented:
Note for reviewers/testers: Only the last commit is part of this PR, the other 2 commits are part of PR 64889 which is a prereq

NitzanMordhai reviewed (changes requested):
Bill that looks like a really good idea!
please check my comments

bill-scales commented:
jenkins test make check

bill-scales commented:
jenkins test make check

bill-scales commented:
jenkins test make check arm64

bill-scales commented:
jenkins test make check arm64

bill-scales commented:
jenkins test make check arm64

bill-scales committed:
qa: Improve failure reporting for watchdog timeouts

A daemon such as an osd or mon crashing can lead to teuthology tests
failing, these failures are detected by the daemon watchdog. Currently
the failure message just indicates that a watchdog timeout occurred.

Add an AssertTracker class that processes the stderr output from
daemon processes, taking insipration from teuthology scrape.py to
capture the assert string and backtrace if the daemon crashes

Enhance the watchdog to report the assertion and backtrace if
a daemon has crashed so this gets reported as the teuthoolgy
job failure reason

Add WatchedProcess for radosbench

Signed-off-by: Bill Scales <bill_scales@uk.ibm.com>

bill-scales commented:
Ready for review:

* First pass review comments from Nitzan have been addressed.
* First pass review comments from Patrick have been addressed including changing from daemon-helper to stdin-killer and some refactoring to move more of the code into watched_process.py.
* Both rados.py and radosbench.py now use watched_process and clean up properly when daemons fail during the test

Here are two teuthology test runs - one with ceph_test_rados and one with rados bench where I deliberately killed a daemon during the test to show that the watchdog timeout fires, the I/O exerciser gets stopped and the test exits reporting which daemon crashed and providing a stack backtrace about the crash (not particularly interesting in this test as it was an external signal):

https://pulpito.ceph.com/billscales-2025-12-10_08:36:32-rados:thrash-erasure-code-main-distro-default-smithi/
https://pulpito.ceph.com/billscales-2025-12-10_10:45:32-rados:thrash-erasure-code-main-distro-default-smithi/

harriscr commented on qa/tasks/ceph.py:
```diff
@@ -1401,6 +1402,69 @@ def run_daemon(ctx, config, type_):
+                return Exception(self.assertion)
+            return None
+
+        def get_failure_time(self):
```
The way this code  written right now you don't even need this method, you can just use <class>.failure_reason and access the internal variable directly. This does allow both read and write access though, so the self.failure_reason can be overwritten from outside the class.

If you wanted to keep failure reason private to the class then it should be ```self._failure_reason``` in the init, and then create a property method to allow read access externally e.g.
```
@property
def failure_time(self):
    return self._failure_time
```
This would be read-access only, and prevent the internal variable from being changed by an external process

harriscr commented on qa/tasks/ceph.py:
```diff
@@ -1401,6 +1402,69 @@ def run_daemon(ctx, config, type_):
+        def get_failure_time(self):
+            return self.failure_time
+
+        def match_id(self, role, id_):
```
probably a personal preference thing, but as this matches both id and role I would make the function name something like 'match_role_and_id'

harriscr commented on qa/tasks/watched_process.py:
```diff
@@ -4,42 +4,59 @@
 """
+import logging
 
 from abc import ABCMeta, abstractmethod
```
we're not using abstractmethod any more, so it can be removed

harriscr reviewed (approved):
Looks good to me
